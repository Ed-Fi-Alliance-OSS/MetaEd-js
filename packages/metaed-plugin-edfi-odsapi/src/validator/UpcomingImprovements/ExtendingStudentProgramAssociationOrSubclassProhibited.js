// @flow

// 2.2.X.1 - METAED-701 - ODS-827
import type { MetaEdEnvironment, ValidationFailure, TopLevelEntity } from 'metaed-core';

function isStudentProgramAssociationOrSubclass(topLevelEntity: TopLevelEntity): boolean {
  if (topLevelEntity.metaEdName === 'StudentProgramAssociation') return true;
  if (topLevelEntity.type !== 'associationSubclass') return false;
  if (topLevelEntity.baseEntity) return isStudentProgramAssociationOrSubclass(topLevelEntity.baseEntity);
  return false;
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.associationExtension.forEach(associationExtension => {
    if (!associationExtension.baseEntity) return;
    if (isStudentProgramAssociationOrSubclass(associationExtension.baseEntity)) {
      failures.push({
        validatorName: 'ExtendingStudentProgramAssociationOrSubclassProhibited',
        category: 'warning',
        message: `[ODS-827] ${associationExtension.typeHumanizedName} ${
          associationExtension.metaEdName
        } is an extension of StudentProgramAssociation or its subclass.  The ODS/API does not currently support this pattern and will fail to build.`,
        sourceMap: associationExtension.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
