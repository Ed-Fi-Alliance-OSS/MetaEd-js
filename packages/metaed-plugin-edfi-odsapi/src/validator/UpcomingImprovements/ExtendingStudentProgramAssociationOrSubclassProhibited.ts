// 2.2.X.1 - METAED-701 - ODS-827
import { versionSatisfies, V2Only, getAllEntitiesOfType } from '@edfi/metaed-core';
import {
  MetaEdEnvironment,
  ValidationFailure,
  TopLevelEntity,
  PluginEnvironment,
  SemVer,
  AssociationExtension,
} from '@edfi/metaed-core';

function isStudentProgramAssociationOrSubclass(topLevelEntity: TopLevelEntity): boolean {
  if (topLevelEntity.metaEdName === 'StudentProgramAssociation') return true;
  if (topLevelEntity.type !== 'associationSubclass') return false;
  if (topLevelEntity.baseEntity) return isStudentProgramAssociationOrSubclass(topLevelEntity.baseEntity);
  return false;
}

const validatorName = 'ExtendingStudentProgramAssociationOrSubclassProhibited';
const targetTechnologyVersion: SemVer = V2Only;

function isTargetTechnologyVersion(metaEd: MetaEdEnvironment): boolean {
  return versionSatisfies(
    (metaEd.plugin.get('edfiOdsApi') as PluginEnvironment).targetTechnologyVersion,
    targetTechnologyVersion,
  );
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  if (!isTargetTechnologyVersion(metaEd)) return failures;

  (getAllEntitiesOfType(metaEd, 'associationExtension') as AssociationExtension[]).forEach(
    (associationExtension: AssociationExtension) => {
      if (!associationExtension.baseEntity) return;
      if (isStudentProgramAssociationOrSubclass(associationExtension.baseEntity)) {
        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-827] ${associationExtension.typeHumanizedName} ${associationExtension.metaEdName} is an extension of StudentProgramAssociation or its subclass.  The ODS/API does not currently support this pattern and will fail to build.`,
          sourceMap: associationExtension.sourceMap.metaEdName,
          fileMap: null,
        });
      }
    },
  );

  return failures;
}
