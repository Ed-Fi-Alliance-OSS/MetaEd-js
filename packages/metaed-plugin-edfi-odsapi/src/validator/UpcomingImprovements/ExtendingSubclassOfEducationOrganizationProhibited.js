// @flow

// 2.1.X.1 - METAED-701
import type { MetaEdEnvironment, ValidationFailure, TopLevelEntity } from 'metaed-core';

function isEducationOrganizationSubclass(topLevelEntity: TopLevelEntity): boolean {
  if (topLevelEntity.type !== 'domainEntitySubclass') return false;
  if (topLevelEntity.baseEntityName === 'EducationOrganization') return true;
  if (topLevelEntity.baseEntity) return isEducationOrganizationSubclass(topLevelEntity.baseEntity);
  return false;
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.domainEntityExtension.forEach(domainEntityExtension => {
    if (!domainEntityExtension.baseEntity) return;
    if (isEducationOrganizationSubclass(domainEntityExtension.baseEntity)) {
      failures.push({
        validatorName: 'ExtendingSubclassOfEducationOrganizationProhibited',
        category: 'warning',
        message: `${domainEntityExtension.typeHumanizedName} ${domainEntityExtension.metaEdName} is an extension of an EducationOrganization subclass.  The ODS/API does not currently support this pattern and will fail to build.`,
        sourceMap: domainEntityExtension.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
