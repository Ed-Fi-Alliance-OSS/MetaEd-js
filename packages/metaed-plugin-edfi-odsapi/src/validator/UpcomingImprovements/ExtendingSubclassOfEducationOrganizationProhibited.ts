// 2.1.X.1 - METAED-701 - ODS-827
import { versionSatisfies, V2Only, getAllEntitiesOfType } from 'metaed-core';
import {
  MetaEdEnvironment,
  ValidationFailure,
  TopLevelEntity,
  PluginEnvironment,
  SemVer,
  DomainEntityExtension,
} from 'metaed-core';

function isEducationOrganizationSubclass(topLevelEntity: TopLevelEntity): boolean {
  if (topLevelEntity.type !== 'domainEntitySubclass') return false;
  if (topLevelEntity.baseEntityName === 'EducationOrganization') return true;
  if (topLevelEntity.baseEntity) return isEducationOrganizationSubclass(topLevelEntity.baseEntity);
  return false;
}

const validatorName = 'ExtendingSubclassOfEducationOrganizationProhibited';
const targetTechnologyVersion: SemVer = V2Only;

function isTargetTechnologyVersion(metaEd: MetaEdEnvironment): boolean {
  return versionSatisfies(
    (metaEd.plugin.get('edfiOdsApi') as PluginEnvironment).targetTechnologyVersion,
    targetTechnologyVersion,
  );
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  if (!isTargetTechnologyVersion(metaEd)) return failures;

  (getAllEntitiesOfType(metaEd, 'domainEntityExtension') as DomainEntityExtension[]).forEach(
    (domainEntityExtension: DomainEntityExtension) => {
      if (!domainEntityExtension.baseEntity) return;
      if (isEducationOrganizationSubclass(domainEntityExtension.baseEntity)) {
        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-827] ${domainEntityExtension.typeHumanizedName} ${domainEntityExtension.metaEdName} is an extension of an EducationOrganization subclass.  The ODS/API does not currently support this pattern and will fail to build.`,
          sourceMap: domainEntityExtension.sourceMap.metaEdName,
          fileMap: null,
        });
      }
    },
  );

  return failures;
}
