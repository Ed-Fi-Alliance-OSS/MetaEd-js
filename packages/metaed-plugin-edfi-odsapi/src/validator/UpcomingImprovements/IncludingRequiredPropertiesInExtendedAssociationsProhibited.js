// @flow

// 2.2.X.2 - METAED-701
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { getAllEntitiesOfType } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  ((getAllEntitiesOfType(metaEd, 'associationExtension'): any): Array<AssociationExtension>).forEach(
    (associationExtension: AssociationExtension) => {
      associationExtension.properties.forEach(property => {
        if (property.isRequired) {
          failures.push({
            validatorName: 'IncludingRequiredPropertiesInExtendedAssociationsProhibited',
            category: 'warning',
            message: `${associationExtension.typeHumanizedName} ${
              associationExtension.metaEdName
            } is an extension with required property ${
              property.metaEdName
            }.  The ODS/API does not currently support this pattern.`,
            sourceMap: property.sourceMap.type,
            fileMap: null,
          });
        }
      });
    },
  );

  return failures;
}
