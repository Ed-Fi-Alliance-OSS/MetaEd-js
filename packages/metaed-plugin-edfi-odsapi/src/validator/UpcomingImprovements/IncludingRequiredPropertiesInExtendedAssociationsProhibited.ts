// 2.2.X.2 - METAED-701
import { MetaEdEnvironment, ValidationFailure, AssociationExtension, SemVer, PluginEnvironment } from 'metaed-core';
import { getAllEntitiesOfType, versionSatisfies } from 'metaed-core';

const targetTechnologyVersion: SemVer = '<3.1';

function isTargetTechnologyVersion(metaEd: MetaEdEnvironment): boolean {
  return versionSatisfies(
    (metaEd.plugin.get('edfiOdsApi') as PluginEnvironment).targetTechnologyVersion,
    targetTechnologyVersion,
  );
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  if (!isTargetTechnologyVersion(metaEd)) return failures;

  (getAllEntitiesOfType(metaEd, 'associationExtension') as Array<AssociationExtension>).forEach(
    (associationExtension: AssociationExtension) => {
      associationExtension.properties.forEach(property => {
        if (property.isRequired) {
          failures.push({
            validatorName: 'IncludingRequiredPropertiesInExtendedAssociationsProhibited',
            category: 'warning',
            message: `${associationExtension.typeHumanizedName} ${
              associationExtension.metaEdName
            } is an extension with required property ${
              property.metaEdName
            }.  The ODS/API does not currently support this pattern.`,
            sourceMap: property.sourceMap.metaEdName,
            fileMap: null,
          });
        }
      });
    },
  );

  return failures;
}
