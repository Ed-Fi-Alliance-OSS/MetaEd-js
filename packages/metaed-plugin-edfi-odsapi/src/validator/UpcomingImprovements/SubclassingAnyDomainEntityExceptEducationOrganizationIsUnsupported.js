// @flow

// 3.1.X.12 - METAED-701
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.domainEntitySubclass.forEach(domainEntitySubclass => {
    if (!domainEntitySubclass.baseEntity) return;
    if (!domainEntitySubclass.namespaceInfo.isExtension) return;
    if (domainEntitySubclass.baseEntityName !== 'EducationOrganization') {
      failures.push({
        validatorName: 'SubclassingAnyDomainEntityExceptEducationOrganizationIsUnsupported',
        category: 'warning',
        message: `${domainEntitySubclass.typeHumanizedName} ${
          domainEntitySubclass.metaEdName
        } is not an EducationOrganization subclass.  EducationOrganization subclasses are the only Domain Entity subclasses currently supported by the ODS/API.`,
        sourceMap: domainEntitySubclass.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
