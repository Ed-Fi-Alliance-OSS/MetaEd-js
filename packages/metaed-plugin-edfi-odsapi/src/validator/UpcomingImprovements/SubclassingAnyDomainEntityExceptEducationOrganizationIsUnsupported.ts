// 3.1.X.12 - METAED-701
// METAED-761 - disabled the following validators:
//    SubclassingAssessmentFamilyIsUnsupported - 3.1.X.7 - METAED-701 - ODS-1324
//    SubclassingAssessmentIsUnsupported - 3.1.X.6 - METAED-701 - ODS-1324
//    SubclassingLearningObjectiveIsUnsupported - 3.1.X.8 - METAED-701 - ODS-1324
//    SubclassingPostSecondaryEventIsUnsupported - 3.1.X.9 - METAED-701 - ODS-1324
//    SubclassingStudentAcademicRecordIsUnsupported - 3.1.X.10 - METAED-701 - ODS-1324
//    SubclassingStudentIsUnsupported - 3.1.X.11 - METAED-701 - ODS-1324 - ODS-1183
//    SubclassingParentIsUnsupported - 3.1.X.13 - METAED-701 - ODS-1183
//    SubclassingStaffIsUnsupported - 3.1.X.14 - METAED-701 - ODS-1183
import { MetaEdEnvironment, ValidationFailure, DomainEntitySubclass } from 'metaed-core';
import { getAllEntitiesOfType } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  (getAllEntitiesOfType(metaEd, 'domainEntitySubclass') as Array<DomainEntitySubclass>).forEach(
    (domainEntitySubclass: DomainEntitySubclass) => {
      if (!domainEntitySubclass.baseEntity) return;
      if (!domainEntitySubclass.namespace.isExtension) return;
      if (domainEntitySubclass.baseEntityName !== 'EducationOrganization') {
        failures.push({
          validatorName: 'SubclassingAnyDomainEntityExceptEducationOrganizationIsUnsupported',
          category: 'error',
          message: `${domainEntitySubclass.typeHumanizedName} ${
            domainEntitySubclass.metaEdName
          } is not an EducationOrganization subclass.  EducationOrganization subclasses are the only Domain Entity subclasses currently supported by the ODS/API.`,
          sourceMap: domainEntitySubclass.sourceMap.type,
          fileMap: null,
        });
      }
    },
  );

  return failures;
}
