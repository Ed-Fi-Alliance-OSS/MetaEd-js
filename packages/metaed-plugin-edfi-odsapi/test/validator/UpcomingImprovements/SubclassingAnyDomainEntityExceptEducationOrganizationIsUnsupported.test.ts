// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  DomainEntityBuilder,
  DomainEntitySubclassBuilder,
  NamespaceBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/UpcomingImprovements/SubclassingAnyDomainEntityExceptEducationOrganizationIsUnsupported';

describe('when a domain entity subclass subclasses EducationOrganization', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const baseEntityName = 'EducationOrganization';
  const subclassName = 'SubclassName';
  let failures: ValidationFailure[];

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(baseEntityName)
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName1', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'EXTENSION')
      .withStartDomainEntitySubclass(subclassName, baseEntityName)
      .withBooleanProperty('PropertyName2', 'doc', true, false)
      .withEndDomainEntitySubclass()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new DomainEntitySubclassBuilder(metaEd, []));

    const coreNamespace: Namespace | undefined = metaEd.namespace.get('EdFi');
    if (coreNamespace == null) throw new Error();
    const extensionNamespace: Namespace | undefined = metaEd.namespace.get('Extension');
    if (extensionNamespace == null) throw new Error();
    extensionNamespace.dependencies.push(coreNamespace);

    const entity = coreNamespace.entity.domainEntity.get(baseEntityName);
    const subclass = extensionNamespace.entity.domainEntitySubclass.get(subclassName);

    if (entity && subclass) subclass.baseEntity = entity;
    failures = validate(metaEd);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when a domain entity subclass subclasses a non-EducationOrganization domain entity', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const baseEntityName = 'NotEducationOrganization';
  const subclassName = 'SubclassName';
  let failures: ValidationFailure[];

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(baseEntityName)
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName1', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'EXTENSION')
      .withStartDomainEntitySubclass(subclassName, baseEntityName)
      .withBooleanProperty('PropertyName2', 'doc', true, false)
      .withEndDomainEntitySubclass()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new DomainEntitySubclassBuilder(metaEd, []));

    const coreNamespace: Namespace | undefined = metaEd.namespace.get('EdFi');
    if (coreNamespace == null) throw new Error();
    const extensionNamespace: Namespace | undefined = metaEd.namespace.get('Extension');
    if (extensionNamespace == null) throw new Error();
    extensionNamespace.dependencies.push(coreNamespace);

    const entity = coreNamespace.entity.domainEntity.get(baseEntityName);
    const subclass = extensionNamespace.entity.domainEntitySubclass.get(subclassName);

    if (entity && subclass) subclass.baseEntity = entity;

    failures = validate(metaEd);
  });

  it('should have validation failures', (): void => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('SubclassingAnyDomainEntityExceptEducationOrganizationIsUnsupported');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot(
      'when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failure -> sourceMap',
    );
  });
});
