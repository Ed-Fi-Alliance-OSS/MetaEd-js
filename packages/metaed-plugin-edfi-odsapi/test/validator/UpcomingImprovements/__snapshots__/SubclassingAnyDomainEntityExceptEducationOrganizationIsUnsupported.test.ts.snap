// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failures: when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failure -> message 1`] = `"Domain Entity Subclass SubclassName is not an EducationOrganization subclass.  EducationOrganization subclasses are the only Domain Entity subclasses currently supported by the ODS/API."`;

exports[`when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failures: when a domain entity subclass subclasses a non-EducationOrganization domain entity should have validation failure -> sourceMap 1`] = `
Object {
  "column": 16,
  "line": 11,
  "tokenText": "SubclassName",
}
`;
